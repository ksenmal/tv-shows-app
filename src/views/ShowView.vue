<script setup>
import { ref, onMounted } from "vue";
import { useRoute } from "vue-router";
import ActorsList from "../components/ActorsList.vue";
import LoaderSpinner from "../components/LoaderSpinner.vue";
import api from "../utils/api";
import fn from "../utils/functions.js";

const show = ref(null);
const isLoading = ref(false);
const route = useRoute();

onMounted(() => {
  isLoading.value = true;
  api
    .getShow(route.params.id)
    .then((res) => {
      isLoading.value = false;
      show.value = res;
    })
    .catch((error) => {
      isLoading.value = false;
      console.error("getShow error: ", error);
    });
});
</script>

<template>
  <LoaderSpinner v-if="isLoading" />
  <p v-else-if="!show" class="no-show-msg">Show cannot be found</p>
  <main v-else>
    <div class="stats clearfix">
      <img
        v-if="show.image && show.image.medium"
        :src="show.image.medium"
        alt=""
      />
      <h2>
        {{ show.name }}
        <span v-if="show.rating" class="rating">{{ show.rating.average }}</span
        ><span class="star">â˜…</span>
      </h2>
      <p class="years">
        ( {{ fn.getYear(show.premiered) }} - {{ fn.getYear(show.ended) }} )
      </p>
      <div class="genres">
        <span v-for="genre in show.genres" :key="genre">{{ genre }}</span>
      </div>
      <p v-if="show.network && show.network.country">
        Country: {{ show.network.country.name }}
      </p>
      <p v-if="show.network">
        Platform:
        <a
          :href="show.network.officialSite"
          target="_blank"
          rel="noopener"
          class="external-link"
          >{{ show.network.name }}</a
        >
      </p>
      <p v-if="show.officialSite">
        Show's official site:
        <a
          :href="show.officialSite"
          target="_blank"
          rel="noopener"
          class="external-link"
          >{{ show.officialSite }}</a
        >
      </p>
    </div>
    <p class="summary">{{ fn.clearHTMLtags(show.summary) }}</p>

    <div
      v-if="show._embedded && show._embedded.cast && show._embedded.cast.length"
      class="cast"
    >
      <h3>Starring:</h3>
      <ActorsList :list="show._embedded.cast" />
    </div>
  </main>
</template>

<style scoped>
.clearfix::after {
  content: "";
  clear: both;
  display: table;
}
.external-link::after {
  content: "ðŸ¡­";
}
main {
  margin: 1rem;
}
h2 {
  font-size: 2rem;
}
.years {
  font-size: 1.5rem;
}
.stats {
  text-align: center;
}
.rating {
  color: var(--color-link);
  padding-left: 1.5rem;
}
.star {
  color: var(--color-yellow);
}
.genres {
  margin: 2rem 0;
}
.genres span {
  margin-right: 5px;
  background-color: var(--color-background-grey);
  padding: 5px 10px;
  border-radius: 5px;
}
.summary {
  margin-top: 2rem;
}
.no-show-msg {
  text-align: center;
  margin-top: 2rem;
  font-size: 1.5rem;
}
.cast {
  margin-top: 2rem;
}
.cast h3 {
  text-align: center;
  font-size: 1.5rem;
  margin-bottom: 2rem;
}

@media (min-width: 1024px) {
  main {
    margin: 2rem 4rem;
  }
  .stats {
    text-align: left;
  }
  .stats img {
    float: left;
    margin-right: 2rem;
  }
}
</style>
