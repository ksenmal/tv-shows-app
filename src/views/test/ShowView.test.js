import ShowView from "../ShowView.vue";
import { mount } from "@vue/test-utils";
import { beforeAll, describe, expect, test, vi } from "vitest";

vi.mock("vue-router", () => {
  return {
    useRoute: () => {
      return {
        params: {
          id: 1,
        },
      };
    },
  };
});

vi.mock("../../utils/functions.js", () => {
  return {
    default: {
      clearHTMLtags: vi.fn(),
      getYear: vi.fn(),
    },
  };
});

describe("ShowView suite", () => {
  beforeAll(() => {
    global.fetch = vi.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () =>
          Promise.resolve({
            summary: "Very nice show",
            ended: "2019-02-02",
            image: {
              medium: "https://images.com/img.jpg",
            },
            network: {
              country: {
                name: "US",
              },
            },
            _embedded: {
              cast: [
                {
                  person: {
                    name: "Jon Snow",
                  },
                  character: {
                    image: {
                      medium: "https://images.com/img.jpg",
                    },
                  },
                },
              ],
            },
          }),
      });
    });
  });

  test("ShowView snapshot", () => {
    const wrapper = mount(ShowView);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("ShowView strips HTML tags from summary and gets year from date", async () => {
    const fn = await import("../../utils/functions.js");

    mount(ShowView);
    expect(fn.default.clearHTMLtags).toHaveBeenCalledWith("Very nice show");
    expect(fn.default.getYear).toHaveBeenCalledWith("2019-02-02");
  });
});
