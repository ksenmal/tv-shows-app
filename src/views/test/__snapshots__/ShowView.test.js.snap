// Vitest Snapshot v1

exports[`ShowView suite > ShowView snapshot 1`] = `"<p class=\\"no-show-msg\\" data-v-06c226f2=\\"\\">Show cannot be found</p>"`;
